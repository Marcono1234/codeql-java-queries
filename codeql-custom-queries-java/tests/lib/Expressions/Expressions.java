import java.util.function.Supplier;

class Expressions {
    void test() {
        toString();
        // LocalVariableDeclarationStatement with init
        String s = toString();
        int i;
        i = 0;
        i++;
        new Object();
        // ArrayCreationExpression cannot be a StatementExpression, but a method access
        // on it can be
        new int[] {}.clone();

        // StatementExpressions
        for (int i1 = 0; i1 < 10; i1++) { }
        for (;;) {
            doSomething();
        }

        // Not a StatementExpression
        for (int i2 : new int[] {1}) { }

        switch(1) {
            default -> toString(); // StatementExpression
        }
        // SwitchExpression has no StatementExpression
        String s2 = switch(1) {
            default -> toString();
        };

        // Lambda with non-void return type has no StatementExpression
        Supplier<Object> supplier1 = () -> toString();
        // Lambda with void return type has StatementExpression
        Runnable r = () -> toString();

        // Method reference with non-void return type has no StatementExpression
        Supplier<Object> supplier2 = Expressions::new;
        // Implicit method of method reference contains StatementExpression
        // TODO: Missing, not sure why. Workaround in StmtExpr should cover this
        Runnable r2 = this::toString;
    }

    void doSomething() { }
}