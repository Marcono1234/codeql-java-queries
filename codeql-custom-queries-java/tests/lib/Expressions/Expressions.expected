| Expressions.java:5:9:5:18 | toString(...) |
| Expressions.java:9:9:9:13 | ...=... |
| Expressions.java:10:9:10:11 | ...++ |
| Expressions.java:11:9:11:20 | new Object(...) |
| Expressions.java:14:9:14:28 | clone(...) |
| Expressions.java:17:18:17:23 | i1 |
| Expressions.java:17:35:17:38 | ...++ |
| Expressions.java:19:13:19:25 | doSomething(...) |
| Expressions.java:27:24:27:33 | toString(...) |
| Expressions.java:40:28:40:37 | toString(...) |
| Expressions.java:42:13:42:22 | toString(...) |
| Expressions.java:48:23:48:36 | toString(...) |
